### ISSUE TITLE
RStudio interrupts Shiny applications on close, interrupting onSessionEnded handlers

### ISSUE THREAD

**Issue by kevinushey** [rstudio/rstudio#13394]
### System details

    RStudio Edition : Desktop [Open Source]
    RStudio Version : 2023.8.0.60
    OS Version      : macOS Ventura 13.4.1
    R Version       : R version 4.3.1 (2023-06-16)

### Steps to reproduce the problem

Try running the following Shiny application:

```
ui <- fluidPage("Close me!")
server <- function(input, output, session) {
  onSessionEnded(function() {
    print("before")
    Sys.sleep(5)
    print("after")
    stopApp()
  })
}

shinyApp(ui = ui, server = server)
```

Then, close the RStudio window hosting the Shiny application.

### Describe the problem in detail

The following is printed in the console:

```
[1] "before"
Error: invoke_wrapped: throwing std::runtime_error
```

That is, the "session ended" computation was interrupted, and an error was thrown.

### Describe the behavior you expected

No error.

**Comment by MariaSemple** [comment]
Related to support ticket 91127.

**Comment by ronblum** [comment]
Notes
- This isn't a recent—it occurs in much earlier versions of RStudio. 
- `Sys.sleep(5)` is the culprit.

**Comment by ronblum** [comment]
I see the error in release 1.3.1093 and on, including 1.4.1717. The only releases in which I don't see the error are 1.0.153 and 1.1.463. Version 1.2.5042 does something a little different:

```
Listening on http://127.0.0.1:7848
[1] "before"

Warning: Error in ||: 'length = 9' in coercion to 'logical(1)'
  35: .rs.isSourceCall
  34: FUN
  33: lapply
  32: .rs.recordTraceback
  31: <Anonymous>
  30: Sys.sleep
  29: callback [/Users/ronblum/sleep/app.R#16]
   1: shiny::runApp
```

**Comment by kevinushey** [comment]
The fix should be relatively straightforward after https://github.com/rstudio/shiny/pull/3860 gets merged into shiny, but that PR will probably take a bit of discussion + back and forth until it's ready. If we can get it in soon, then we could consider it, I think.

**Comment by ronblum** [comment]
CC @jthomasmock

**Comment by jthomasmock** [comment]
I put this into Product Backlog for now as that's a rough error message. Are we blocked by the Shiny PR upstream or do we need help from that team?

**Comment by kevinushey** [comment]
Yes, I believe we need the Shiny team to take that PR (or something similar) for us to be able to support this.

**Comment by wch** [comment]
A workaround that users can do now is to wrap their callback code in `suspendInterrupts()`.

Here's a demonstration of what `suspendInterrupts()` does.

First, we'll define a function called `spinloop()`, which does basically the same thing as `Sys.sleep()`. I'll explain why later.

```R
# Use this function instead of Sys.sleep()
spinloop <- function(t) {
  t_stop <- as.numeric(Sys.time()) + t
  while (Sys.time() < t_stop) { NULL }
}
```

Then if you run this block of code, and the press Ctrl-C right away, it will print `"Begin"`, but it will not print `"End"`:

```R
{
  print("Begin")
  spinloop(3)
  print("End")
}
#> [1] "Begin"
# Press Ctrl-C to send an interrupt now -- it will not print "End"
```


However, if you wrap it with `suspendInterrupts()`, the Ctrl-C will not interrupt that block of code, and it **will** print `"End"`:
 
```R
suspendInterrupts({
  print("Begin")
  spinloop(3)
  print("End")
})
#> [1] "Begin"
# Press Ctrl-C to send an interrupt now
#> [1] "End"
```

*****

Note that if you use `Sys.sleep()` instead of `spinloop()`, it will be interruptable. This is by design -- `Sys.sleep()` explicitly makes itself interruptable even if it is inside of `suspendInterrupts()`. See https://bugs.r-project.org/show_bug.cgi?id=18673 for more information.

However, for real-world use cases in Shiny, the code in the `onSessionEnded` callback generally won't make itself interruptable, so for the sake of demonstration, I'll use the `spinloop()` function.

The example below is based on the original, except it uses `spinloop()` instead of `Sys.sleep()` (to better simulate cleanup code), and it wraps the code in the callback inside of `suspendInterrupts()`.

```R
spinloop <- function(t) {
  t_stop <- as.numeric(Sys.time()) + t
  while (Sys.time() < t_stop) { NULL }
}

ui <- fluidPage("Close me!")
server <- function(input, output, session) {
  onSessionEnded(function() {
    suspendInterrupts({
      print("before")
      spinloop(5)
      print("after")
      stopApp()
    })
  })
}

shinyApp(ui = ui, server = server)
```

With this example, when the user closes the window, it prints both `"before"` and `"after"`.

**Comment by ronblum** [comment]
Verified in
- RStudio Desktop 2024.04.0-daily+495 on MacOS 14.4
- Posit Workbench 2024.04.0-daily+478.pro12 on Ubuntu 22.04 via Chrome on MacOS 14.4

Checked
- Using the example above
- Adding the definition of the server in the default Shiny web app created in RStudio
- Both single- and multiple-file apps
